### PR TITLE
Fix task failure signal race condition

### PR DESCRIPTION
**Pull request by pettyjamesm**
## Description
Fixes the sequence of operations around `TaskResource` checking for task failure until after the buffer response future is completed. Otherwise, tasks could inappropriately announce completion rather than failure to consumers which can yield incorrect query results. 

## Additional context and related issues

Discovered while attempting to debug a swath of test failures producing incorrect results in https://github.com/trinodb/trino/pull/15478 which made the sequence of interleaving operations necessary to trigger the correctness bug more likely to occur.

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# General
* Fix a race condition that could allow queries to indicate normal completion instead of failure and produce incorrect results.
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java [hunk 1]
```diff
@@ -778,4 +768,37 @@ private void failStuckSplitTasks()
...
+
+        public String getTaskInstanceId()
+        {
+            return task.getTaskInstanceId();
+        }
+
+        public boolean isTaskFailed()
+        {
+            TaskState state = task.getTaskState();
+            return (state == TaskState.ABORTED | state == TaskState.FAILED);
```
Comment (linzebing) [inline core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java:796]:
```suggestion
            return (state == TaskState.ABORTED || state == TaskState.FAILED);
```

Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java]: Expanded to a `switch` instead since future work will consider more states as "failed".

File: core/trino-main/src/main/java/io/trino/server/TaskResource.java [hunk 2]
```diff
@@ -189,11 +189,14 @@ public void getTaskInfo(
...
 
         Duration waitTime = randomizeWaitTime(maxWait);
-        ListenableFuture<TaskInfo> futureTaskInfo = addTimeout(
-                taskManager.getTaskInfo(taskId, currentVersion),
-                () -> taskManager.getTaskInfo(taskId),
-                waitTime,
-                timeoutExecutor);
+        ListenableFuture<TaskInfo> futureTaskInfo = taskManager.getTaskInfo(taskId, currentVersion);
+        if (!futureTaskInfo.isDone()) {
+            futureTaskInfo = addTimeout(
```
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/server/TaskResource.java:194]: nit: maybe extract improvements into separate commit / commits

File: core/trino-main/src/main/java/io/trino/server/TaskResource.java [hunk 3]
```diff
@@ -516,4 +496,31 @@ private static Duration randomizeWaitTime(Duration waitTime)
...
+            status = Status.OK;
+        }
+
+        return Response.status(status)
+                .entity(entity)
+                .header(TRINO_TASK_INSTANCE_ID, result.getTaskInstanceId())
+                .header(TRINO_PAGE_TOKEN, result.getToken())
+                .header(TRINO_PAGE_NEXT_TOKEN, result.getNextToken())
+                .header(TRINO_BUFFER_COMPLETE, result.isBufferComplete())
+                .header(TRINO_TASK_FAILED, taskWithResults.isTaskFailed())
```
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/server/TaskResource.java:523]: nit: maybe add a comment here, something along the line "check task failure after getting the result"

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java [hunk 4]
```diff
@@ -163,17 +163,22 @@ public SqlTaskExecution(
...
+            if (taskHandle == null) {
+                return;
+            }
             // The scheduleDriversForTaskLifeCycle method calls enqueueDriverSplitRunner, which registers a callback with access to this object.
-            // The call back is accessed from another thread, so this code cannot be placed in the constructor.
+            // The call back is accessed from another thread, so this code cannot be placed in the constructor. This must also happen before outputBuffer
+            // callbacks are registered to prevent a task completion check before task lifecycle splits are created
             scheduleDriversForTaskLifeCycle();
+            // Output buffer state change listener callback must not run in the constructor to avoid leaking a reference to "this" across to another thread
+            outputBuffer.addStateChangeListener(new CheckTaskCompletionOnBufferFinish(SqlTaskExecution.this));
```
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java:181]: nit: move to a separate commit

File: core/trino-main/src/main/java/io/trino/execution/buffer/LazyOutputBuffer.java [hunk 5]
```diff
@@ -209,7 +208,7 @@ public ListenableFuture<BufferResult> get(OutputBufferId bufferId, long token, D
         if (outputBuffer == null) {
             synchronized (this) {
                 if (delegate == null) {
-                    if (stateMachine.getState() == FINISHED) {
+                    if (stateMachine.getState().isTerminal()) {
```
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/execution/buffer/LazyOutputBuffer.java:211]: We should not return `complete` when the buffer is not `FINISHED`

File: core/trino-main/src/main/java/io/trino/execution/SqlTask.java [hunk 6]
```diff
@@ -167,19 +171,22 @@ private void initialize(Consumer<SqlTask> onDone, CounterStat failedTasks)
             }
 
             // store final task info
-            while (true) {
+            taskHolderLock.lock();
```
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/execution/SqlTask.java:174]: nit: consider `synchronized(this)` and reorder the `taskExecution.start();` as discussed offline (unless we missed something that can break the flow)

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java (hunk 4 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java:181]: Actually, this is all related code. The movement to here alone would cause issues with `scheduleDriversWithTaskLifecycle()` not being called when `taskHandle == null`. This is "fine" because it doesn't break the task state machine model today, but causes problems in a world with terminating task states.

File: core/trino-main/src/main/java/io/trino/execution/buffer/LazyOutputBuffer.java (hunk 5 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/buffer/LazyOutputBuffer.java:211]: You are correct, fixing in next revision.

File: core/trino-main/src/main/java/io/trino/server/TaskResource.java [hunk 7]
```diff
@@ -322,52 +328,26 @@ public void getResults(
             return;
         }
 
-        TaskState state = taskManager.getTaskStatus(taskId).getState();
-        boolean taskFailed = state == TaskState.ABORTED || state == TaskState.FAILED;
-
         long start = System.nanoTime();
-        ListenableFuture<BufferResult> bufferResultFuture = taskManager.getTaskResults(taskId, bufferId, token, maxSize);
+        SqlTaskManager.SqlTaskWithResults taskWithResults = taskManager.getTaskResults(taskId, bufferId, token, maxSize);
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/server/TaskResource.java]: Done

File: core/trino-main/src/main/java/io/trino/server/TaskResource.java (hunk 2 above)
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/server/TaskResource.java:195]: Done

**Review by arhimondr — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/execution/buffer/LazyOutputBuffer.java [hunk 8]
```diff
@@ -310,19 +312,31 @@ public void destroy()
...
         if (outputBuffer == null) {
             synchronized (this) {
                 if (delegate == null) {
                     // ignore abort if the buffer already in a terminal state.
-                    stateMachine.abort();
+                    if (!stateMachine.abort()) {
+                        return;
+                    }
 
-                    // Do not free readers on fail
```
Comment (pettyjamesm) [inline core/trino-main/src/main/java/io/trino/execution/buffer/LazyOutputBuffer.java:320]: The assumption is that sending `emptyResults(taskInstanceId, 0, completed: false)` is safe, even though the previous comment suggested otherwise, because `TaskResource` will respond similarly when the timeout threshold is reached. Sending these responses instead of letting them time out should propagate task failure information more quickly than allowing the timeout to expire.
